Asynchronous mode

dgant opened the pull request:
This pull request adds an optional asynchronous operation mode to JBWAPI, similar to the non-blocking mode of TorchCraft.

The goal is to allow bot authors to more reliably ensure real-time performance for their bots. Async mode runs in two threads: One for communication with BWAPI, and one for bot operation. This allows bot code to run independently of the BWAPI control flow, and for JBWAPI to release its lock-step control of StarCraft in a timely fashion even if bot code is still running.

When a new frame from BWAPI is ready, JBWAPI copies it into a queue of frames for the bot to process. JBWAPI then waits 40ms (configurable) for the bot to finish processing all the frames in the queue. When the bot has done so or time has elapsed, JBWAPI returns control to BWAPI.

dgant committed:
Added client configuration object to replace individual constructor arguments.

dgant committed:
Extracted portions of Client to GameDataUtils. Removed the bot loop entirely for reinsertion via BotWrapper, so this revision is in a non-working state.

dgant committed:
Restored existing (sync) behavior.

dgant committed:
Finished first pass async implementation, untested

dgant committed:
Assigning the asynchronous bot perspective to Game

dgant committed:
Corrected copying of shared memory to FrameBuffer. Corrected measurement of bot idleness and frame buffer size.

dgant committed:
First end-to-end working version. Added side effect queue to route commands to live shared memory instead of uselessly writing them to the frame buffer.

dgant committed:
Fixed strings

dgant committed:
Replaced sleeps with awaits

dgant committed:
Was getting memory allocation exceptions when replacing FrameBuffer's direct ByteBuffers on successive games, due either to fragmentation or leaking (or both). Now reusing the FrameBuffer across games so we only have to allocate once, for the first game. Replaced some unnecessary references to ClientData with GameData. Cleaned up BWClient game loop.

dgant committed:
Fixed end-of-game behavior. Moved synchronization out of BWClient and removed 'idleness' concept from BotWrapper.

dgant committed:
Implemented frame timeouts.

dgant committed:
Reduced stdout spamming

dgant committed:
Fixed unit test which relied on ClientData being non-final for mocking

dgant committed:
Removed commented-out printlns. Stubbed diagnostics configuration variable.

dgant committed:
Fixed erroneous conversion between nanoseconds and milliseconds. First pass at performance metrics.

dgant committed:
Fixed some collection and display of performance metrics

dgant committed:
Fixed condition for waiting for bot to finish (instead of returning control immediately always). Improved some metrics

JasperGeurtz committed:
use memcpy to copy the framebuffer

dgant committed:
[Tests failing] Finished test environment for sync/async tests. Configured tests to run with sufficient memory to allocate framebuffers.

dgant committed:
[Failing tests] Unit tests now terminate even if the bot thread dies due to assertion errors

dgant committed:
Restored stepwise logging to monitor asynchronous steps

dgant committed:
Fixed bug in which BotWrapper was looking at bot data to determine the live frame

dgant committed:
Removed console noise caused by bot thread dying (as expected) from exception test

dgant committed:
All tests passing. Stubbed out performance metrics tests and added bot frame performance test

dgant committed:
Fixing performance metrics. Fixed framebuffer thinking it's full one frame too soon.

dgant committed:
Finished asynchronous unit tests; fixed miscellaneous performance measurements

dgant committed:
Cleanup

dgant committed:
Fixed unit tests

dgant commented on pom.xml:
```diff
@@ -63,6 +63,8 @@
                     <parallel>all</parallel> <!-- Run tests in parallel-->
                     <useUnlimitedThreads>true</useUnlimitedThreads>
+                    <argLine>-Xms1g</argLine>
+                    <argLine>-Xmx1g</argLine>
```
Ensures unit tests have enough memory to run, especially in parallel.
A BWAPI frame takes 31MB, and thus asynchronous tests require N * 31MB free contiguous memory blocks to run.
The synchronization test environment will check that sufficient memory is available at runtime.

dgant commented on src/main/java/bwapi/BWClient.java:
```diff
@@ -7,58 +7,112 @@
         Objects.requireNonNull(eventListener);
-        this.debugConnection = debugConnection;
         this.eventListener = eventListener;
     }
```
debugConnection moved to BWClientConfiguration

dgant commented on src/main/java/bwapi/BWClient.java:
```diff
@@ -7,58 +7,112 @@
      *
      * @param autoContinue automatically continue playing the next game(s). false by default
      */
+    @Deprecated
```
This overload is replaced by one which takes a BWClientConfiguration to indicate thhe autoContinue setting.

dgant commented on src/main/java/bwapi/Client.java:
```diff
@@ -258,67 +250,17 @@ void update(final EventHandler handler) {
-
-    String eventString(final int s) {
-        return gameData.getEventStrings(s);
-    }
```
The logic for applying changes to the game has moved to Game and SideEffect/SideEffectQueue

dgant commented on src/main/java/bwapi/ClientData.java:
```diff
@@ -1,1993 +1,2000 @@
-package bwapi;
```
Autogenerated file. I probably changed line endings when I regenerated it. Give a 👍 if you'd like to see the line endings kept the same. The changes are more clearly highlighted in DumpToClient.java

dgant committed:
Minor tweaks

JasperGeurtz commented on src/main/java/bwapi/BWClientConfiguration.java:
```diff
@@ -0,0 +1,77 @@
+/**
+ * Configuration for constructing a BWClient
+ */
+public class BWClientConfiguration {
```
I'd prefer the variables to be private and modifiable using a `withDebugConnection()` , with the comment above the variable added to the `withX` method

example API in my mind:
```Java
BWClientConfiguration bwc = new BWClientConfiguration()
    .withAutoContinue()
    .withAsync()
    .withMaxFrameDuration(30);
```

```Java
/**Javadoc*/
public BWClientConfiguration withMaxFrameDuration(int value) {
    if (value < 1) {
        throw new IllegalArgumentException("maxFrameDuration needs to be a non-negative number (it's how long JBWAPI waits for a bot response before returning control to BWAPI).");
    }
    this.maxFrameDuration = value;
    return this;
}
```

This way we don't need a "validate" method as we can check the arguments on creating the configuration
and users don't need to change internal variable names of objects.

dgant replied:
Will do, although I think the validate method is still necessary since it considers interactions between properties and it would be annoying to require them to be set in a specific order, and it doesn't prevent modification of the properties.

JasperGeurtz commented on src/main/java/bwapi/BWClient.java:
```diff
@@ -7,58 +7,112 @@
+     * @return The number of frames between the one exposed to the bot and the most recent received by JBWAPI.
+     * This tracks the size of the frame buffer except when the game is paused (which results in multiple frames arriving with the same count).
+     */
+    public int framesBehind() {
```
Iis this tested in the synchro test? impl looks good to me, but just making sure

dgant replied:
Yes, there are a few tests which verify its values.

JasperGeurtz commented on src/main/java/bwapi/BWClientConfiguration.java:
```diff
@@ -0,0 +1,77 @@
+        }
+    }
+
+    public void log(String value) {
```
feels kind of weird to have this in the config but I get it, I would make the method package-private though

JasperGeurtz commented on src/main/java/bwapi/BotWrapper.java:
```diff
@@ -0,0 +1,191 @@
+            gameOver = gameOver || gameData.getEvents(i).getType() == EventType.MatchEnd;
+        }
+
+        if (configuration.async) {
```
for example if we keep `async` as a public variable, a user could change this on frame 10 or something, and I don't know if it's desirable to be able to swithc over to async on the fly (e.g. what side effects that could have)

JasperGeurtz commented on src/main/java/bwapi/EventHandler.java:
```diff
@@ -30,10 +18,10 @@ public void operation(final ClientData.Event event) {
             //case 3: //MenuFrame
             case SendText:
-                eventListener.onSendText(client.eventString(event.getV1()));
+                eventListener.onSendText(game.clientData().gameData().getEventStrings(event.getV1()));
```
maybe we should add a `game.gameData()` method that calls `clientData().gameData()` internally to shorten some code/avoid duplication

dgant replied:
It's an option. I found myself mostly deleting methods like that because it was really easy to confuse the live game data with the potentially-delayed bot data. The fewer methods which returned one of those, the easier. Could go either way on it.

JasperGeurtz replied:
ok, then maybe the method names should be more descriptive for the potentially-delayed bot data calls? But I'm also fine with leaving it like this

dgant replied:
Renaming Game.clientData() to "botClientData()" and Client.clientData() to "liveClientData()" to distinguish

JasperGeurtz commented on src/main/java/bwapi/FrameBuffer.java:
```diff
@@ -0,0 +1,173 @@
+                }
+            }
+        }
+        catch(Exception ignored) {}
```
we should probably print the exception in config.log() here

JasperGeurtz commented on src/main/java/bwapi/GameDataUtils.java:
```diff
@@ -0,0 +1,79 @@
+/*
```
Don't need to put the LICENSE here, only one file from Hannes was used, which was licensed under MIT, which is still there for those reasons, for the other files the repo wide MIT license should be valid.

If not, I'll automate to add to all files later, but that doesn't have to be done for this PR

dgant replied:
I think I included it here because this code was copy-pasted from that file. Will double check.

dgant replied:
Confirmed: this came from https://github.com/JavaBWAPI/JBWAPI/blob/develop/src/main/java/bwapi/Client.java#L267 which is under Hannes' license.

JasperGeurtz commented on src/main/java/bwapi/Game.java:
```diff
@@ -100,14 +99,18 @@
     private Text.Size textSize = Text.Size.Default;
     private boolean latcom = true;
 
+    public final SideEffectQueue sideEffects = new SideEffectQueue();
```
sorry if I'm going over this too quickly, does this have an impact on non-async JBWAPI?

dgant replied:
The side effect queue is still used, but it's flushed immediately so the behavior is identical. It introduces a lambda allocation and function call per command issued, which I could remove with some additional work.

JasperGeurtz replied:
It would really surprise me if commands issued would ever be the bottleneck of `jbwapi`'s side, could you maybe just add what you wrote as a comment above the variable for future reference?

dgant replied:
It just occurred to me that if latcom fabricates game state based on commands issued in the current frame, then this implementation breaks latcom, as the queue is flushed all at once after onFrame rather than on the fly. Will address.

dgant replied:
Async probably doesn't play nice with latcom out of the box. Latcom likely needs to differentiate between when a command was actually issued, and when the bot attempted to issue it. I haven't thought this through (and am out of touch on how latcom works). This needs some more attention, even if the initial solution is "don't allow latcom with async"

JasperGeurtz replied:
We can just make latcom and async mutually exclusive, e.g. by enabling async, latcom gets disabled.

We can also add a "withLatCom(boolean enabled)" flag to the BWConfig, same for UserInput. So we don't need to call these in onStart anymore

JasperGeurtz commented on src/main/java/bwapi/PerformanceMetric.java:
```diff
@@ -0,0 +1,121 @@
+    private final String name;
+    private final long maxAllowed;
+
+    public double minValue = Long.MAX_VALUE;
```
I would make these private with `getters`

JasperGeurtz commented on src/main/java/bwapi/PerformanceMetrics.java:
```diff
@@ -0,0 +1,95 @@
+     * Likely to be at least a little bit of an undercount,
+     * given that the tournament module is timing a superset of JBWAPI's execution time.
+     */
+    public PerformanceMetric totalFrameDuration;
```
private with `getters`

JasperGeurtz commented on src/test/java/DumpToClient.java:
```diff
@@ -92,8 +92,15 @@ public static void main(String[] args) throws IOException {
+            out.println("    GameData gameData() {");
+            out.println("        return gameData;");
+            out.println("    }");
+            out.println("    public void setBuffer(ByteBuffer buffer) {");
```
does `setBuffer` need to be public?

JasperGeurtz reviewed (commented):
PR LGTM design wise, can't really test it performance wise myself right now.

Some remarks on the `public`/`private`ness of the variables.

Thanks a lot for the PR!

JasperGeurtz commented on src/main/java/bwapi/BotWrapper.java:
```diff
@@ -0,0 +1,191 @@
+        if (configuration.async) {
+            configuration.log("Main: onFrame asynchronous start");
+            long startNanos = System.nanoTime();
+            long endNanos = startNanos + configuration.maxFrameDurationMs * 1000000;
```
```Java
long endNanos = startNanos + (long)configuration.maxFrameDurationMs * 1000000;
```
to fix the LGTM warning

dgant commented:
Should refer to https://github.com/JavaBWAPI/JBWAPI/pull/63/files for JDK 9+ compatibility

dgant committed:
Clarified thresholds for performance metrics

dgant committed:
Implemented unsafe asynchronous mode, allowing the bot to read directly from live game data while the frame buffer is being populated. There's some additional cruft lying around and there's no guarantee that this will actually work in practice, but the test case is passing.

dgant committed:
Implemented partial frame copies, cutting about 3/4 off the amount of data we need to copy per frame

dgant committed:
Added some more metrics. Added multiple thresholds to metrics.

dgant committed:
Trying various tactics to resolve timeouts. Added a timer resolution hack which should force Windows to use the lowest resolution (16ms) for GetTickCount() (though it appeared to be doing that in all environments anyway).  Reduced thread priority for JBWAPI + bot threads to increase likelihood that BWAPI receives priority sooner.

dgant committed:
Removed unnecessary allocations in Game API.

dgant committed:
Made thread priority reduction dependent on async mode

dgant committed:
Added performance metric for frame duration which includes time to write to shared memory.

dgant committed:
Assembled complete suite of frame-duration and BWAPI communication performance metrics

dgant committed:
Fixed typo in performance logging

dgant committed:
Added performance metrics related to event count to diagnose whether 4.4 tournament module bugs are ongoing

dgant committed:
Fixed up unit tests

dgant committed:
Moved PerformanceMetric(s) properties behind getters

dgant committed:
Put configuration properties behind getters and chain-setters

dgant committed:
Configuration now can not be modified after game starts

dgant committed:
Removed the attempted TimerResolutionThread hack

dgant committed:
Cosmetic changes/cleanup

dgant committed:
Merged jdk9 fixes into async branch

dgant committed:
Fixed use case of copying frames via ByteBuffer API

dgant committed:
Disabled latency compensation in async mode.

dgant commented:
New in the pull request:
- Addressed the comments left on the original pull request
- Introduced unsafe asynchronous mode, for eking out extra performance. Unsafe mode runs the bot against live game data while copying the live game data to the frame buffer in the background. It's "unsafe" because there is no active mechanism preventing modification of which buffer the bot is reading from simultaneous with a bot's attempt to read from it. In practice I have never observed this *failing* -- PurpleWave has been happily running with this mode on the AIIDE ladder and BASIL.
- More granular performance measurements. Includes metrics which I found useful for diagnosing performance on the AIIDE ladder.
- Fixes synchronous latency compensation mode. Since we observed that async mode and latcom are incompatible, I added checks to prevent inadvertently enabling both. These modes could be reconciled in the future.

I believe all loose ends are tied up and this PR is ready for review in its current state.

dgant committed:
Made configuration accessors public

JasperGeurtz committed:
remove parallel tests and replace memcpy with unsafe.copyMemory

dgant committed:
Reenabled memory arguments for unit tests

dgant committed:
Removed more Windows-specific JNA calls

dgant committed:
Added test re-runs. Made classes final to aid inlining

dgant committed:
Merge remote-tracking branch 'origin/develop' into async

dgant committed:
De-finalized classes to fix mocking. Widened threshold on synchronization tests

dgant committed:
Trying to ensure tests are run with adequate memory available

dgant commented:
We're on passing tests now and merge compatibility with develop. @JasperGeurtz what do you think?

JasperGeurtz commented:
LGTM, thank you for finishing this up!

three questions:
 - ~is the `<rerunFailingTestsCount>10</rerunFailingTestsCount>`  argument needed in the pom.xml, seems to run fine on my PC without?~ timed tests are always a bit flaky
 - ~are the frames in the framebuffer being cleared up properly? (I remember a while ago this to be a potential problem, but don't know what the end result was)~ Only mattered for tests as we reallocate every test
 - ~I also want to add the "linux openbw" branch by  @Bytekeeper  for a "2.0" release. How (in)compatible are the two branches? They both modify how BWClient etc. work, so there will be some conflicts at least. ANy thoughts @Bytekeeper @dgant~ Will be merged after, so not an issue

I don't mind version 2.0 being a bit experimental for now, and just keeping 1.5.1 as the "stable" branch that anyone can use
